File: scripts/install-skill.mjs
#!/usr/bin/env node
import { cpSync, existsSync, mkdirSync, statSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const RUNTIME_PATHS = [
  "SKILL.md",
  "agents",
  "locales",
  "prompts",
  "schemas",
  "templates",
  "validators",
];

function usage() {
  return `Usage:
  npx svgflow-response-skill [--skills-dir <path>] [--source <path>] [--name <folder>]
  npx github:fanchou/svgflow_response_skill [--skills-dir <path>]

Options:
  --skills-dir <path>  Target skills directory. Defaults to $CODEX_HOME/skills or ~/.codex/skills.
  --source <path>      Source folder containing SKILL.md. Defaults to this package root.
  --name <folder>      Installed skill folder name. Defaults to svgflow-response.
  --help               Show this help.
`;
}

function parseArgs(argv) {
  const args = {
    name: "svgflow-response",
    skillsDir: undefined,
    source: undefined,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      args.help = true;
    } else if (arg === "--skills-dir") {
      args.skillsDir = argv[++index];
    } else if (arg === "--source") {
      args.source = argv[++index];
    } else if (arg === "--name") {
      args.name = argv[++index];
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }

  return args;
}

function homeDir() {
  return process.env.HOME || process.env.USERPROFILE;
}

function expandHome(path) {
  if (!path || path === "~") return homeDir();
  if (path.startsWith("~/") || path.startsWith("~\\")) {
    return join(homeDir(), path.slice(2));
  }
  return path;
}

function defaultSkillsDir() {
  if (process.env.SVGFLOW_SKILLS_DIR) return process.env.SVGFLOW_SKILLS_DIR;
  if (process.env.SKILLS_DIR) return process.env.SKILLS_DIR;
  if (process.env.CODEX_HOME) return join(process.env.CODEX_HOME, "skills");
  return join(homeDir(), ".codex", "skills");
}

function packageRoot() {
  return resolve(dirname(fileURLToPath(import.meta.url)), "..");
}

function copyRuntime(sourceRoot, targetRoot) {
  if (!existsSync(join(sourceRoot, "SKILL.md"))) {
    throw new Error(`SKILL.md not found in source folder: ${sourceRoot}`);
  }

  mkdirSync(targetRoot, { recursive: true });

  for (const relativePath of RUNTIME_PATHS) {
    const source = join(sourceRoot, relativePath);
    if (!existsSync(source)) continue;
    const target = join(targetRoot, relativePath);
    const isDirectory = statSync(source).isDirectory();
    cpSync(source, target, {
      recursive: isDirectory,
      force: true,
      errorOnExist: false,
    });
  }
}

try {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    process.stdout.write(usage());
    process.exit(0);
  }

  const sourceRoot = resolve(expandHome(args.source || packageRoot()));
  const skillsDir = resolve(expandHome(args.skillsDir || defaultSkillsDir()));
  const targetRoot = join(skillsDir, args.name);

  copyRuntime(sourceRoot, targetRoot);

  process.stdout.write(`Installed SVGFlow skill to: ${targetRoot}\n`);
  process.stdout.write("Restart your agent or start a new session so it can rediscover skill metadata.\n");
} catch (error) {
  process.stderr.write(`${error.message}\n\n${usage()}`);
  process.exit(1);
}

File: scripts/install_skill.py
#!/usr/bin/env python3
"""Install the runtime SVGFlow skill folder into a Codex skills directory."""

from __future__ import annotations

import argparse
import os
import shutil
from pathlib import Path


RUNTIME_PATHS = [
    "SKILL.md",
    "agents",
    "locales",
    "prompts",
    "schemas",
    "templates",
    "validators",
    "scripts/render_infographic.py",
]


def default_skills_dir() -> Path:
    codex_home = os.environ.get("CODEX_HOME")
    if codex_home:
        return Path(codex_home).expanduser() / "skills"
    return Path.home() / ".codex" / "skills"


def copy_path(source: Path, target: Path) -> None:
    if source.is_dir():
        shutil.copytree(source, target, dirs_exist_ok=True)
    else:
        target.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(source, target)


def install(source_root: Path, skills_dir: Path, name: str) -> Path:
    if not (source_root / "SKILL.md").is_file():
        raise SystemExit(f"SKILL.md not found in source root: {source_root}")

    target_root = skills_dir / name
    target_root.mkdir(parents=True, exist_ok=True)

    for relative in RUNTIME_PATHS:
        source = source_root / relative
        if source.exists():
            copy_path(source, target_root / relative)

    return target_root


def main() -> None:
    parser = argparse.ArgumentParser(description="Install SVGFlow as a Codex skill.")
    parser.add_argument(
        "--source",
        default=".",
        help="Path to the repository or skill source root containing SKILL.md.",
    )
    parser.add_argument(
        "--skills-dir",
        default=None,
        help="Codex skills directory. Defaults to $CODEX_HOME/skills or ~/.codex/skills.",
    )
    parser.add_argument(
        "--name",
        default="svgflow-response",
        help="Installed skill folder name.",
    )
    args = parser.parse_args()

    source_root = Path(args.source).expanduser().resolve()
    skills_dir = Path(args.skills_dir).expanduser().resolve() if args.skills_dir else default_skills_dir()
    target_root = install(source_root, skills_dir, args.name)
    print(f"Installed SVGFlow skill to: {target_root}")
    print("Restart Codex or start a new thread so the skill metadata can be rediscovered.")


if __name__ == "__main__":
    main()
